### PR TITLE
Update task to error if defined enttiy update / create fails

### PR DESCRIPTION
**Pull request by Anirudh9794**
Signed-off-by: Aniruddha Shamasundar <aniruddha.9794@gmail.com>

To help us process your pull request efficiently, please include: 

* When defined entity create/update fails in the sync portion of cluster operations, the task will be left in Running state forever
* Fix by adding exception handling to update tasks if defined entity create/ update fails

Testing done:
* create (with and without rights)
* resize (with and without rights)
* delete (with and without rights)
* upgrade (with and without rights)

@rocknes @sahithi @sakthisunda

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774)

### CODE REVIEW

**Comment by ChintanShahVMware** [comment]
___
*[container_service_extension/def_/cluster_service.py, line 194 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-0alaBrB53Xjq8zdH:-MJ-0alaBrB53Xjq8zdI:b-65s3ka) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L194)):*
> ```Python
>                 tenant_org_context=org_context)
>             def_entity = self.entity_svc.get_native_entity_by_name(cluster_name)  # noqa: E501
>         except Exception as err:
> ```

Hmm Is there a way you can be more specific about the exception? 
See if you can use a specific exception instead of generic one.

**Comment by ChintanShahVMware** [comment]
___
*[container_service_extension/def_/cluster_service.py, line 269 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-0t6iATouehOCEX8D:-MJ-0t6iATouehOCEX8E:benpsw9) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L269)):*
> ```Python
>         try:
>             curr_entity = self.entity_svc.update_entity(cluster_id, curr_entity)  # noqa: E501
>         except Exception as err:
> ```

Same comment as above.

**Comment by ChintanShahVMware** [comment]
___
*[container_service_extension/def_/cluster_service.py, line 316 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-1-nZ0D_jZ86J1Ggz:-MJ-1-nZ0D_jZ86J1Gh-:benpsw9) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L316)):*
> ```Python
>             # lets vCD authorize the user for delete operation.
>             self.entity_svc.delete_entity(cluster_id)
>         except Exception as err:
> ```

Same comment as above.

**Comment by ChintanShahVMware** [comment]
___
*[container_service_extension/def_/cluster_service.py, line 316 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-1-nZ0D_jZ86J1Ggz:-MJ-17-87u102q3JFDtl:bf38p2b) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L316)):*
<details><summary><i>Previously, ChintanShahVMware wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Basically all the exceptions you are catching..

**Review by sakthisunda — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774)* status: all files reviewed, 3 unresolved discussions (waiting on @Anirudh9794)

**Review by rocknes — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774)* status: all files reviewed, 7 unresolved discussions (waiting on @Anirudh9794 and @ChintanShahVMware)

___
*[container_service_extension/def_/cluster_service.py, line 316 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-1-nZ0D_jZ86J1Ggz:-MJ-Q2uy1Q0x949imSX-:bxpkjy7) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L316)):*
<details><summary><i>Previously, ChintanShahVMware wrote…</i></summary><blockquote>

Basically all the exceptions you are catching..
</blockquote></details>

I think the idea is to catch any and all exception.
___
*[container_service_extension/def_/cluster_service.py, line 184 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-PWrV3kRuTBibzUW-:-MJ-PWrW3QZjZtHN6XQQ:b-dnlwg8) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L184)):*
> ```Python
>                                              org_name=def_entity.entity.metadata.org_name)  # noqa: E501
>             org_context = org_resource.href.split('/')[-1]
>         msg = f"Creating cluster '{cluster_name}' ({def_entity.id}) " \
> ```

Wouldn't def_entity.id be None at this point?
___
*[container_service_extension/def_/cluster_service.py, line 196 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-PkIF14p6_2z8joLC:-MJ-PkIG3SMj2OOaZIWc:bor2rqv) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L196)):*
> ```Python
>         except Exception as err:
>             self._update_task(vcd_client.TaskStatus.ERROR,
>                               message=msg,
> ```

msg is still "creating cluster..." right? it should be the cause of error, or creation failed etc.
___
*[container_service_extension/def_/cluster_service.py, line 1111 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-QMNQ5ZBaafNRKuyZ:-MJ-QMNQ5ZBaafNRKuy_:b26vrfu) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L1111)):*
> ```Python
>             msg = f"Unexpected error while upgrading cluster " \
>                   f"'{cluster_name}'"
>             LOGGER.error(f"{msg}: {err}", exc_info=True)
> ```

if exec_info is set to true, the exception is captured as part of the log message, no need to add it explicitly.
___
*[container_service_extension/def_/cluster_service.py, line 1276 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-QaOrDnABTSYUUtmv:-MJ-QaOs8Y53X-Pt5EEG:blsrgcy) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L1276)):*
> ```Python
>             self.entity_svc.resolve_entity(cluster_id)
>         except Exception as err:
>             msg = f"Failed to resolfe defined entity for cluster {cluster_id}"
> ```

nit: resolve

**Review by Anirudh9794 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774)* status: all files reviewed, 6 unresolved discussions (waiting on @ChintanShahVMware and @rocknes)

___
*[container_service_extension/def_/cluster_service.py, line 194 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-0alaBrB53Xjq8zdH:-MJ-Xwqk0VhVVFSxD22s:b-543m1s) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L194)):*
<details><summary><i>Previously, ChintanShahVMware wrote…</i></summary><blockquote>

Hmm Is there a way you can be more specific about the exception? 
See if you can use a specific exception instead of generic one.
</blockquote></details>

Since any exception should lead to task getting updated, I think all exceptions should be caught.
___
*[container_service_extension/def_/cluster_service.py, line 269 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-0t6iATouehOCEX8D:-MJ-ZAnU6y3ZP2zQD2b2:bubqkpi) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L269)):*
<details><summary><i>Previously, ChintanShahVMware wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Since it is better to catch all Exception, I will leave this as is.
___
*[container_service_extension/def_/cluster_service.py, line 316 at r1](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-1-nZ0D_jZ86J1Ggz:-MJ-ZQ9sB0fxOznjGx4I:bgpf4ac) ([raw file](https://github.com/vmware/container-service-extension/blob/95a70aa5a3af662299dcb9b54202842353a3c5fc/container_service_extension/def_/cluster_service.py#L316)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I think the idea is to catch any and all exception.
</blockquote></details>

Since it is better to catch all Exception, I will leave this as is.
___
*[container_service_extension/def_/cluster_service.py, line 184 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-PWrV3kRuTBibzUW-:-MJ-XmiI-Q6nh8dMfK88:b59o6m3) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L184)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Wouldn't def_entity.id be None at this point?
</blockquote></details>

Done. removed` def_entity.id`
___
*[container_service_extension/def_/cluster_service.py, line 196 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-PkIF14p6_2z8joLC:-MJ-Y9_IEGzS7FuTtNPR:b-s950pb) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L196)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

msg is still "creating cluster..." right? it should be the cause of error, or creation failed etc.
</blockquote></details>

I was changing this with UI perspective, it will show the task name as the message and there is another column to indicate the status of the task. I will add an error message here instead of that.
___
*[container_service_extension/def_/cluster_service.py, line 1111 at r2](https://reviewable.io/reviews/vmware/container-service-extension/774#-MJ-QMNQ5ZBaafNRKuyZ:-MJ-_GRtB6VP1lXNYYv0:b-896fix) ([raw file](https://github.com/vmware/container-service-extension/blob/9cf100a82e3b19ab08b357e0e6cc080044d78daf/container_service_extension/def_/cluster_service.py#L1111)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

if exec_info is set to true, the exception is captured as part of the log message, no need to add it explicitly.
</blockquote></details>

Done.

**Review by rocknes — APPROVED** [review]
:lgtm:

Can you please also make similar changes in vcdbroker that might lead to empty error messages?

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774#-:-MJ-c8mjCwtqUYsMzONa:b-qysuun)* status: all files reviewed, 3 unresolved discussions (waiting on @ChintanShahVMware)

**Review by rocknes — APPROVED** [review]
A different PR for vcdbroker is absolutely fine :)

> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774#-:-MJ0635a43mykYX4jLcD:b92a32t)* status: all files reviewed, 3 unresolved discussions (waiting on @ChintanShahVMware)

**Review by sahithi — APPROVED** [review]
:lgtm: Please test thoroughly.

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774#-:-MJ0Hbme33jpcYGhcwxJ:b308qn0)* status: all files reviewed, 3 unresolved discussions (waiting on @ChintanShahVMware)

**Review by sahithi — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/vmware/container-service-extension/774)* status: all files reviewed, 3 unresolved discussions (waiting on @ChintanShahVMware)
